doc/reference/configuration/configuration_reference.rst

```diff
@@ -2877,6 +2878,120 @@ This section describes configuration parameters related to roles.
+
+    There are two types of application roles:
+
+    -   Built-in roles. For example, the `CRUD <https://github.com/tarantool/crud>`__ module provides the ``roles.crud-storage`` and ``roles.crud-router`` roles that enable CRUD operations in a sharded cluster.
+    -   Custom roles. For example, you create a custom role to define a stored procedure or implement a supplementary service, such as an e-mail notifier or a replicator.
+
+    To create a custom role, you need to define the following methods and fields:
+
+    -   ``validate()``
+
+        Validate a role's configuration.
+        This method is called when the initial configuration is applied or the configuration is reloaded for the instance with this role.
+        ``validate()`` should throw an error if the validation fails.
+
+    -   ``apply()``
+
+        Apply a role's configuration.
+        This method is called when the initial configuration is applied or the configuration is reloaded for the instance with this role.
+        ``apply()`` should throw an error if the specified configuration can't be applied.
+
+    -   ``stop()``
+
+        Stop a role.
+        This method is called on configuration reload if the role is removed from ``roles`` for the given instance.
+        ``stop()`` should cancel all background fibers and clean up used resources.
+        If these actions can't be performed, ``stop()`` should throw an error.
+
+    -   (Optional) ``dependencies``
+
+        Define a role's dependencies as an array.
+        For roles that depend on each other, their ``validate()`` and ``apply()`` methods are executed taking into account the dependencies.
+        Suppose, there are three roles that depend on each other as follows:
+
+        ..  code-block:: none
+
+            role1
+                └─── role2
+                         └─── role3
+
+        In this case, ``validate()`` and ``apply()`` for these roles are executed in the following order:
+
+        ..  code-block:: none
+
+            role3 -> role2 -> role1
+
+
+    **Example: Built-in role**
+
+    The examples below show how to enable the router and storage-specific roles provided by the `CRUD <https://github.com/tarantool/crud>`__ module:
+
+    -   ``roles.crud-router``:
+
+        ..  literalinclude:: /code_snippets/snippets/sharding/instances.enabled/sharded_cluster_crud/config.yaml
+            :language: yaml
+            :start-at: routers
+            :end-at: roles.crud-router
+            :dedent:
+
+    -   ``roles.crud-storage``:
+
+        ..  literalinclude:: /code_snippets/snippets/sharding/instances.enabled/sharded_cluster_crud/config.yaml
+            :language: yaml
+            :start-at: storages
+            :end-at: roles.crud-storage
+            :dedent:
+
+    You can find the full example here: `sharded_cluster_crud <https://github.com/tarantool/doc/tree/latest/doc/code_snippets/snippets/sharding/instances.enabled/sharded_cluster_crud>`_.
+
+    **Example: Custom role**
```

Comment:
Questions that I was asked by early adopters:

* Where to place an initialization code (cartridge's `<role>.init()`)?[^1]
* How to create a space?[^2]

It is beyond the reference scope, but these questions are important and we definitely should highlight them in examples/recipes/howtos.

[^1]: On the top level of a module. One may wrap it to a function and call on the top level:

      ```lua
      local M = {}

      local function init()
          <...>
      end

      function M.validate(cfg)
          <...>
      end

      function M.apply(cfg)
          <...>
      end

      function M.stop()
          <...>
      end

      init()

      return M
      ```
[^2]: We should do it on an RW instance (using a fiber plus `box.ctl.wait_rw()` or using `box.watch('box.status', <...>)` or using `box.ctl.on_election()`, but the latter is applicable only for `replication.failover: election`).

      Note: `apply()` is not invoked if an instance goes to RW as part of a Raft election or appointed by a failover coordinator. Unlike cartridge, there is no `is_leader` property in the `apply` function, its purpose is different. If some actions should be performed, when an instance goes to RW, a code should catch **this** event, not a reconfiguration.

      Note 2: But the schema initialization code (that creates a space) should be called from `apply` too, because a role may be enabled, when an instance is already RW.

      We should create a space once, so `if_not_exists` should be used (or `box.once`, but there were some problems around it with timeouts and error handling).